packages/mail-client/src/mail-client.ts

```diff
@@ -140,20 +139,57 @@ async function createSocket(
       port: mailDestination.port!
     }
   };
-  const socketConnection =
-    await SocksClient.createConnection(connectionOptions);
-
-  const socksSocket = socketConnection.socket as SocksSocket;
-  // Setting `_readableListening` to true in the next line makes the socket readable.
-  // Otherwise nodemailer is not able to receive the SMTP greeting message and send emails.
-  socksSocket._readableState.readableListening = true;
-  return socksSocket;
+  const { socket } = await SocksClient.createConnection(connectionOptions);
+
+  // Workaround for incorrect order of events in nodemailer https://github.com/nodemailer/nodemailer/issues/1684
+  await resendGreetingUntilReceived(socket);
+
+  return socket;
+}
+
+function retrieveGreeting(socket: Socket): Promise<Buffer> {
+  logger.debug('Waiting for SMTP greeting message...');
+  return new Promise((resolve, reject) => {
+    const onData = data => {
+      logger.debug(`Data received from mail socket: ${data?.toString()}`);
+      if (data?.toString().startsWith('220')) {
+        logger.debug('Removing mail socket listeners...');
+        socket.removeListener('data', onData);
+        socket.removeListener('error', onError);
+        resolve(data);
+      }
+    };
+
+    const onError = err => {
+      reject(new Error(err));
+    };
+
+    socket.on('data', onData);
+    socket.on('error', onError);
+  });
+}
+
+async function resendGreetingUntilReceived(socket: Socket): Promise<void> {
+  const greeting = await retrieveGreeting(socket);
+
+  retry(
+    () => {
+      // resend the greeting message until a listener is attached
+      // note: this is dangerous because there could be another listener that is not the mailer
+      if (socket.listenerCount('data')) {
+        socket.emit('data', greeting);
+        return;
+      }
```

Comment:
Wait, I am not sure but don't you need to throw some error here if no listener found? Otherwise will retry actually work?

Reply:
I think since the return value of `socket.emit()` also tells you if there is a listener or not, can't we write

```ts
if (!socket.emit('data', greeting)) {
  throw new Error('Failed to re-emit event. No data listener found.');
}
```